pkg/query-service/app/traces/v3/query_builder.go

```diff
@@ -0,0 +1,385 @@
+	v3.FilterOperatorContains:        "ILIKE",
+	v3.FilterOperatorNotContains:     "NOT ILIKE",
+	v3.FilterOperatorExists:          "has(%s%s, '%s')",
+	v3.FilterOperatorNotExists:       "NOT has(%s%s, '%s')",
+}
+
+func getColumnName(key v3.AttributeKey, keys map[string]v3.AttributeKey) (string, error) {
+	key, err := enrichKeyWithMetadata(key, keys)
+	if err != nil {
+		return "", err
+	}
+	if key.IsColumn {
+		return key.Key, nil
+	}
+	filterType, filterDataType := getClickhouseTracesColumnDataTypeAndType(key)
+	return fmt.Sprintf("%s%s['%s']", filterDataType, filterType, key.Key), nil
+}
+
+func getClickhouseTracesColumnDataTypeAndType(key v3.AttributeKey) (v3.AttributeKeyType, string) {
+	filterType := key.Type
+	filterDataType := "string"
+	if key.DataType == v3.AttributeKeyDataTypeFloat64 || key.DataType == v3.AttributeKeyDataTypeInt64 {
+		filterDataType = "number"
+	} else if key.DataType == v3.AttributeKeyDataTypeBool {
+		filterDataType = "bool"
+	}
+	if filterType == v3.AttributeKeyTypeTag {
+		filterType = "TagMap"
+	} else {
+		filterType = "resourceTagsMap"
+		filterDataType = ""
+	}
+	return filterType, filterDataType
+}
+
+func enrichKeyWithMetadata(key v3.AttributeKey, keys map[string]v3.AttributeKey) (v3.AttributeKey, error) {
+	if key.Type == "" || key.DataType == "" {
+		// check if the key is present in the keys map
+		if existingKey, ok := keys[key.Key]; ok {
+			key.IsColumn = existingKey.IsColumn
+			key.Type = existingKey.Type
+			key.DataType = existingKey.DataType
+		} else {
+			return key, fmt.Errorf("key not found to enrich metadata")
+		}
+	}
+	return key, nil
+}
+
+// getSelectLabels returns the select labels for the query based on groupBy and aggregateOperator
+func getSelectLabels(aggregatorOperator v3.AggregateOperator, groupBy []v3.AttributeKey, keys map[string]v3.AttributeKey) (string, error) {
+	var selectLabels string
+	if aggregatorOperator == v3.AggregateOperatorNoOp {
+		selectLabels = ""
+	} else {
+		for _, tag := range groupBy {
+			filterName, err := getColumnName(tag, keys)
+			if err != nil {
+				return "", err
+			}
+			selectLabels += fmt.Sprintf(", %s as `%s`", filterName, tag.Key)
+		}
+	}
+	return selectLabels, nil
+}
+
+// getZerosForEpochNano returns the number of zeros to be appended to the epoch time for converting it to nanoseconds
+func getZerosForEpochNano(epoch int64) int64 {
+	count := 0
+	if epoch == 0 {
+		count = 1
+	} else {
+		for epoch != 0 {
+			epoch /= 10
+			count++
+		}
+	}
+	return int64(math.Pow(10, float64(19-count)))
+}
+
+func buildTracesFilterQuery(fs *v3.FilterSet, keys map[string]v3.AttributeKey) (string, error) {
+	var conditions []string
+
+	if fs != nil && len(fs.Items) != 0 {
+		for _, item := range fs.Items {
+			toFormat := item.Value
+			// generate the key
+			columnName, err := getColumnName(item.Key, keys)
+			if err != nil {
+				return "", err
+			}
+			var fmtVal string
+			if toFormat != "" {
+				fmtVal = utils.ClickHouseFormattedValue(toFormat)
+			}
```

Comment:
`ClickHouseFormattedValue` takes care of empty values. Does the `buildTracesFilterQuery` a valid SQL if the value is empty sting? i.e `some_key=''`

Reply:
Added a test case for empty string